src/backend/distributed/shared_library_init.c

```diff
@@ -2146,12 +2167,69 @@ StatisticsCollectionGucCheckHook(bool *newval, void **extra, GucSource source)
 static void
 CitusAuthHook(Port *port, int status)
 {
+	if (!IsCitusInternalBackend())
+	{
+		/*
+		 * We raise the shared connection counter pre-emptively. As a result, we may
+		 * have scenarios in which a few simultaneous connection attempts prevent
+		 * each other from succeeding, but we avoid scenarios where we oversubscribe
+		 * the system.
+		 *
+		 * By also calling RegisterClientBackendCounterDecrement here, we immediately
+		 * lower the counter if we throw a FATAL error below. The client connection
+		 * counter may temporarily exceed maxClientConnections in between.
+		 */
+		RegisterClientBackendCounterDecrement();
+		IncrementClientBackendCounter();
```

Comment:
Should we call the functions `*ExternalClients*` as we are currently only counting those?

The only existing caller already only need this information (verified that the semantics still holds with this change).

Reply:
makes sense